src/NexusMods.DataModel/NxArchiveManager.cs

```diff
@@ -185,30 +184,206 @@ public async Task ExtractFiles(IEnumerable<(Hash Src, AbsolutePath Dest)> files,
+                {
+                    // This mode is only supported if start of decompressed data is at offset 0 of decompressed buffer.
+                    // If this is unsupported (rarely in this hot path) we go back to 'slow' approach.
+                    canFastDecompress = false;
+                    goto fallback;
+                }
+
+                // Get block index.
+                var blockIndexOffset = extractable.BlockIndex - _entry.FirstBlockIndex;
+                var start = (long)chunkSize * blockIndexOffset;
+                var decompSizeInChunk = _entry.DecompressedSize - (ulong)start;
+                var length = Math.Min((long)decompSizeInChunk, chunkSize);
+
+                fixed (byte* compressedPtr = compressedBlock)
+                fixed (byte* ptr = buffer)
+                {
+                    Compression.Decompress(method, compressedPtr, blockSize, ptr, (int)length);
+                }
+
+                return;
+            }
+
+            // This is the logic in case of multiple outputs, e.g. if user specifies an Array + File output.
+            // It incurs additional memory copies, which may bottleneck when extraction is done purely in RAM.
+            // Decompress the needed bytes.
+            using var extractedBlock = MemoryPool<byte>.Shared.Rent(extractable.DecompressSize);
+
+            fixed (byte* compressedPtr = compressedBlock)
+            fixed (byte* extractedPtr = extractedBlock.Memory.Span)
+            {
+                // Decompress all.
+                Compression.Decompress(method, compressedPtr, blockSize, extractedPtr,
+                    extractable.DecompressSize);
+
+
+                // Get block index.
+                var blockIndexOffset = extractable.BlockIndex - _entry.FirstBlockIndex;
+                var start = (long)chunkSize * blockIndexOffset;
+                var decompSizeInChunk = _entry.DecompressedSize - (ulong)start;
+                var length = Math.Min((long)decompSizeInChunk, chunkSize);
+
+                fixed (byte* ptr = buffer)
+                {
+                    Buffer.MemoryCopy(extractedPtr + _entry.DecompressedBlockOffset, ptr, length, length);
+                }
+            }
+        }
+
+        private void MakeBlocks(int chunkSize)
+        {
+            // Slow due to copy to stack, but not that big a deal here.
+            var chunkCount = _entry.GetChunkCount(chunkSize);
+            var remainingDecompSize = _entry.DecompressedSize;
+
+            for (var chunkIndex = 0; chunkIndex < chunkCount; chunkIndex++)
+            {
+                var blockIndex = _entry.FirstBlockIndex + chunkIndex;
+                var block = new ExtractableBlock
+                {
+                    BlockIndex = blockIndex,
+                    DecompressSize = _entry.DecompressedBlockOffset + (int)Math.Min(remainingDecompSize, (ulong)chunkSize)
+                };
+
+                _blocks.Add(block);
```

Comment:
This can be preallocated as array rather than using list; since the number of blocks (chunkCount) is known ahead of time.